Apps/AdminWebClient/src/Server/Services/InactiveUserService.cs

```diff
@@ -0,0 +1,195 @@
+                    requestResult.ResourcePayload = inactiveUsers;
+                    requestResult.TotalResultCount = inactiveUsers.Count;
+                }
+            }
+        }
+
+        this.logger.LogDebug("Inactive user count: {Count}...", inactiveUsers.Count);
+        return requestResult;
+    }
+
+    private void UpdateInactiveUser(List<AdminUserProfileView> inactiveUsers, List<UserRepresentation> identityAccessUsers, IdentityAccessRole role)
+    {
+        this.logger.LogDebug("{Role} count: {Count}...", role.ToString(), identityAccessUsers.Count);
+        List<UserRepresentation> users = identityAccessUsers.FindAll(x1 => inactiveUsers.Exists(x2 => x1.Username == x2.Username));
+        this.logger.LogDebug("Filtered matching {Role} count: {Count}...", role.ToString(), users.Count);
+
+        foreach (AdminUserProfileView inactiveUser in inactiveUsers)
+        {
+            UserRepresentation? user = users.Find(x => x.Username == inactiveUser.Username);
+
+            if (user != null)
+            {
+                inactiveUser.UserId = inactiveUser.UserId ??= user.UserId;
+                inactiveUser.Email = inactiveUser.Email ??= user.Email;
+                inactiveUser.FirstName = inactiveUser.FirstName ??= user.FirstName;
+                inactiveUser.LastName = inactiveUser.LastName ??= user.LastName;
+                inactiveUser.RealmRoles = inactiveUser.RealmRoles != null ? (inactiveUser.RealmRoles + ", " + role.ToString()) : role.ToString();
+            }
+        }
+    }
+
+    private void AddInactiveUser(List<AdminUserProfileView> inactiveUsers, List<AdminUserProfile> activeUserProfiles, List<UserRepresentation> identityAccessUsers, IdentityAccessRole role)
+    {
+        this.logger.LogDebug("{Role} count: {Count}...", role.ToString(), identityAccessUsers.Count);
+        List<UserRepresentation> users = identityAccessUsers.FindAll(x1 => !inactiveUsers.Exists(x2 => x1.Username == x2.Username));
+        this.logger.LogDebug("Filtered non-matching {Role} count: {Count}...", role.ToString(), users.Count);
+
+        foreach (UserRepresentation user in users)
+        {
+            AdminUserProfileView? adminUserProfileView = inactiveUsers.Find(x1 => x1.Username == user.Username);
+
+            if (adminUserProfileView == null)
+            {
+                bool adminProfileExists = activeUserProfiles.Exists(x2 => x2.Username == user.Username);
+
+                if (!adminProfileExists)
+                {
+                    AdminUserProfileView adminUserProfile = AdminUserProfileView.FromKeycloakModel(user);
+                    adminUserProfile.RealmRoles = role.ToString();
+                    inactiveUsers.Add(adminUserProfile);
+                }
+            }
+            else
+            {
+                adminUserProfileView.RealmRoles = adminUserProfileView.RealmRoles != null ? (adminUserProfileView.RealmRoles + ", " + role.ToString()) : role.ToString();
+            }
+        }
```

Comment:
Nit: the `Find()` query seems like it will always return `null`, since `users` consists entirely of records that have no corresponding match in the `inactiveUsers` collection.

The loop could probably be simplified by filtering `users` to ones that do not have an active user profile ahead of time.

```
IEnumerable<UserRepresentation> users = identityAccessUsers.Where(x1 =>
    !inactiveUsers.Exists(x2 => x1.Username == x2.Username) &&
    !activeUserProfiles.Exists(x2 => x1.Username == x2.Username));
```

To go a step further, the method _could_ be simplified to a one-liner if the role could be passed into `FromKeycloakModel`:

```
inactiveUsers.AddRange(
    identityAccessUsers.Where(x1 =>
        !inactiveUsers.Exists(x2 => x1.Username == x2.Username) && 
        !activeUserProfiles.Exists(x2 => x1.Username == x2.Username))
    .Select(x => AdminUserProfileView.FromKeycloakModel(x, role.ToString()));
```

Reply:
I updated log messages and provided comments in code to better explain things.